core/src/main/java/com/google/zxing/datamatrix/decoder/DecodedBitStreamParser.java

```diff
@@ -117,15 +126,34 @@ static DecoderResult decode(byte[] bytes) throws FormatException {
     if (resultTrailer.length() > 0) {
       result.append(resultTrailer);
     }
-    return new DecoderResult(bytes, result.toString(), byteSegments.isEmpty() ? null : byteSegments, null);
+    if (isECIencoded) {
+      if (fnc1Positions.containsKey(0) || fnc1Positions.containsKey(4)) {
+        symbologyModifier = 5;
+      } else if (fnc1Positions.containsKey(1) || fnc1Positions.containsKey(5)) {
+        symbologyModifier = 6;
+      } else {
+        symbologyModifier = 4;
+      }
+    } else {
+      if (fnc1Positions.containsKey(0) || fnc1Positions.containsKey(4)) {
+        symbologyModifier = 2;
+      } else if (fnc1Positions.containsKey(1) || fnc1Positions.containsKey(5)) {
+        symbologyModifier = 3;
+      } else {
+        symbologyModifier = 1;
+      }
+    }
+
+    return new DecoderResult(bytes, result.toString(), byteSegments.isEmpty() ? null : byteSegments, null, symbologyModifier);
   }
 
   /**
    * See ISO 16022:2006, 5.2.3 and Annex C, Table C.2
    */
   private static Mode decodeAsciiSegment(BitSource bits,
                                          StringBuilder result,
-                                         StringBuilder resultTrailer) throws FormatException {
+                                         StringBuilder resultTrailer,
+                                         Map fnc1positions) throws FormatException {
```

Comment:
This Map can/should have generic types, no?

Reply:
Good advice.
I'm aware Java has generic types, but it's hard to change habits that have been formed in many years without them and rarely getting to code Java nowadays.